Release/src/http/client/http_client_asio.cpp

```diff
@@ -350,22 +370,374 @@ class asio_client : public _http_client_communicator, public std::enable_shared_
+        {
+            auto connection_ipv6 = client_cast->m_pool.obtain(asio_connection_pool::Type::NEW);
+            invokeUserCallback(connection_ipv6);
+            connect(++m_requestsCount, connection_ipv6, he_endpoints, handler);
+            isIpv6Connecting = static_cast<bool>(connection_ipv6);
+            ++he_endpoints;
+        }
+        
+        if (isValid(he_endpoints) && (he_endpoints)->endpoint().address().is_v4())
+        {
+            auto connection_ipv4 = client_cast->m_pool.obtain(asio_connection_pool::Type::NEW);
+            invokeUserCallback(connection_ipv4);
+            if (isIpv6Connecting)
+            {
+                if (execMode == ExecMode::HE_ENABLED)
+                {
+                    scheduleConnect(++m_requestsCount, connection_ipv4, he_endpoints, handler);
+                }
+            }
+            else
+            {
+                connect(++m_requestsCount, connection_ipv4, he_endpoints, handler);
+            }
+        }
+    }
+    
+    void invokeUserCallback(std::shared_ptr<asio_connection>& connection)
+    {
+        try
+        {
+            if (auto client = m_client.lock())
+            {
+            if (connection->is_ssl())
+            {
+                client->client_config().invoke_nativehandle_options(connection->m_ssl_stream.get());
+            }
+            else
+            {
+                client->client_config().invoke_nativehandle_options(&(connection->m_socket));
+            }
+            }
+        }
+        catch (...)
+        {
+            throw;
+        }
+    }
+    
+    std::string toString(tcp::resolver::iterator endpoints)
+    {
+        std::stringstream ss;
+        
+        auto host = endpoints->host_name();
+        auto service = endpoints->service_name();
+        auto address = endpoints->endpoint().address();
+        auto ip = address.to_string();
+        
+        ss << " host: " << host << ":" << service << " ip: " << ip;
+        
+        return ss.str();
+    }
+    
+    std::string toString(void *id, tcp::resolver::iterator endpoints)
+    {
+        std::stringstream ss;
+        
+        auto host = endpoints->host_name();
+        auto service = endpoints->service_name();
+        auto address = endpoints->endpoint().address();
+        auto ip = address.to_string();
+        
+        ss << "id: " << id <<  toString(endpoints);
+        
+        return ss.str();
+    }
+    
+    void removePendingConnection(std::shared_ptr<asio_connection> connection)
+    {
+        pendingConnections.erase(std::remove(pendingConnections.begin(), pendingConnections.end(), connection), pendingConnections.end());
+    }
+    
+    void handle_tcp_connect(std::weak_ptr<asio_connection> weakConnection, int id, const boost::system::error_code& ec, tcp::resolver::iterator endpoints, ConnectHandler handler)
+    {
+        auto connection = weakConnection.lock();
+        if (!connection)
+        {
+            return;
+        }
```

Comment:
grab a shared ptr to a connection. If not possible means we are already connected and exit.